### PR TITLE
compute,adapter: provide introspection for dataflow error counts

### LINKED ISSUE #21453: Expose dataflow errors in an introspection source
### Feature request

Compute should expose an introspection source that reveals the presence of dataflow errors. Just having a `dataflow_id` to error count mapping would be extremely helpful. We could then use this information in the console to show errored indexes/MVs.

### PR DESCRIPTION
**Pull request by teskje**
This PR adds logging of error counts in dataflow exports to compute, as well as a set of `mz_compute_error_counts` introspection sources/views to surface the logged data. The intent is to provide an easy way for users to find out if any of their dataflows have errors, and for Mz support to be able to monitor error in users' dataflows.

### Motivation

  * This PR adds a known-desirable feature.

Closes #21453

### Tips for reviewer

You can look at the commit separately.

The main tests for these new sources live in `introspection-sources.td`. I tried to think of all the edge cases, but LMK if you find more!

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Add an `mz_internal.mz_compute_error_counts` introspection view that surfaces errors in indexes and materialized views.

### CODE REVIEW

File: src/compute/src/render/mod.rs [hunk 1]
```diff
@@ -526,6 +526,12 @@ where
             Some(ArrangementFlavor::Local(oks, errs)) => {
                 // Set up probes to notify on index frontier advancement.
                 oks.stream.probe_notify_with(probes);
+
+                // Attach logging of dataflow errors.
+                if let Some(logger) = compute_state.compute_logger.clone() {
+                    errs.stream.log_dataflow_errors(logger, idx_id);
```
Comment (antiguru) [inline src/compute/src/render/mod.rs:532]: This will cause all errors to be cloned. Better assign the output to `errs` to avoid attaching multiple operators to the `errs` stream.

File: src/compute/src/logging/compute.rs [hunk 2]
```diff
@@ -398,6 +406,29 @@ impl<A: Allocate> DemuxState<A> {
...
+/// State tracked for each dataflow export.
+struct ExportState {
+    /// The ID of the dataflow maintaining this export.
+    dataflow_id: usize,
+    /// Imports feeding this export, and their frontier delay tracking state.
+    imports: BTreeMap<GlobalId, FrontierDelayState>,
+    /// Number of errors in this export.
+    ///
+    /// We keep this a signed integer to not be tripped up by negative total error counts. Seeing
+    /// negative counts would be a bug, but we still want to handle that gracefully.
```
Comment (antiguru) [inline src/compute/src/logging/compute.rs:432]: I think it's a bug if the _overall_ count across all workers is negative, but the per-worker count is fine to be negative. Only after a keyed exchange, the local counts are guaranteed to be positive.

**Review by antiguru — COMMENTED** [review]
Thanks for picking this line of work! I think it's super important to surface the error counts. I left some comments inline, and the PR looks good. However, I have some architectural questions I'd like to resolve first.

* If I'm not mistaken, we already know the number of errors as the number of records in error arrangements. I'd prefer if we'd re-use this number instead of adding an additional logging path that requires re-reading the arrangement's updates. At the moment, it's hard to associate which arrangement contains errors, and which one doesn't due to the lack of context when we're constructing the Timely operators. But I think there's a way out of this: The arrangement size logging is attached to all arrangements (it's a bug if it's not) and we might be able to specialize the implementation for `RowKeySpine` to only cover `DataflowError`.
* Why do we have to log the counts of errors on `Collection`? All errors should eventually hit an arrangement, or is there a case where this isn't true?
* The last commit seems odd. It seems to me that a better way to surfacing this information is to provide a view that sums up the size of a dataflow including what it depends on, rather then try to squeeze into logging. Indexes are logical abstraction boundaries provided by the users, so I think it makes sense to re-use the boundaries.

**Comment by teskje** [comment]
> If I'm not mistaken, we already know the number of errors as the number of records in error arrangements.

We do only for indexes. In MVs errors don't get arranged, so we don't have arrangements to look at:

```sql
materialize=> create table t (a int);
CREATE TABLE
materialize=> insert into t values (0);
INSERT 0 1
materialize=> create materialized view mv as select 1 / a from t;
CREATE MATERIALIZED VIEW
materialize=> select * from mv;
ERROR:  Evaluation error: division by zero
materialize=> select * from mz_internal.mz_arrangement_sizes;
 operator_id | records | batches | size | capacity | allocations
-------------+---------+---------+------+----------+-------------
(0 rows)
```

(This also applies to SUBSCRIBEs but those get aborted on errors anyway.)

Also, record counts can be very misleading because a) errors are deduplicated and b) removed errors are only reflected after compaction:

```sql
materialize=> create table t (a int);
CREATE TABLE
materialize=> create view v as select 1 / a from t;
CREATE VIEW
materialize=> create default index on v;
CREATE INDEX
materialize=> insert into t values (0), (0), (0);
INSERT 0 3
materialize=> select * from mz_internal.mz_arrangement_sizes;
 operator_id | records | batches | size | capacity | allocations
-------------+---------+---------+------+----------+-------------
 9503602     |       0 |       1 |   16 |       16 |           8
 9503611     |       1 |       1 |   48 |       96 |           3
(2 rows)

materialize=> delete from t;
DELETE 3
materialize=> select * from mz_internal.mz_arrangement_sizes;
 operator_id | records | batches | size | capacity | allocations
-------------+---------+---------+------+----------+-------------
 9503602     |       0 |       1 |   16 |       16 |           8
 9503611     |       2 |       1 |   64 |      120 |           3
(2 rows)
```

(9503611 is the error arrangement here.)

So I think there is no way around reading the updates again, to check their diff. I think this should be fine in the common case where errors are rare. For the uncommon case where there is a huge amount of errors, users will have other issues than slow dataflow performance I assume :)

> Why do we have to log the counts of errors on Collection? All errors should eventually hit an arrangement, or is there a case where this isn't true?

Yes, materialized views (and subscribes).

> The last commit seems odd. It seems to me that a better way to surfacing this information is to provide a view that sums up the size of a dataflow including what it depends on, rather than try to squeeze into logging. Indexes are logical abstraction boundaries provided by the users, so I think it makes sense to re-use the boundaries.

I'm not entirely sure I understand your suggestion here. Do you mean we could write a view that a) somehow figures out when an index is directly reusing another index and b) copies the error counts from the first on in this case? I think that could work, based on that the dependency relationships do show up as expected in `mz_compute_dependencies`:

```sql
materialize=> create table t (a int);
CREATE TABLE
materialize=> create default index on t;
CREATE INDEX
materialize=> create default index on t;
CREATE INDEX
materialize=> create default index on t;
CREATE INDEX
materialize=> create default index on t;
CREATE INDEX
materialize=> select * from mz_internal.mz_compute_dependencies where object_id like 'u%';
 object_id | dependency_id
-----------+---------------
 u39       | u38
 u40       | u39
 u41       | u39
 u42       | u39
(5 rows)
```

The only wrinkle is that the `mz_compute_error_counts_raw` data will not be complete for these reused indexes, which is a bit of a footgun if we ever try to use `_raw` sources for anything. But probably less of an issue than the additional complexity introduced by handling this in logging. I'll give that a try!

File: src/compute/src/render/mod.rs (hunk 1 above)
Comment (teskje) [reply src/compute/src/render/mod.rs:532]: We are not using `errs.stream` anymore (and we can't since we are moving out of it here) so there shouldn't be any cloning happening, right?

File: src/compute/src/logging/compute.rs (hunk 2 above)
Comment (teskje) [reply src/compute/src/logging/compute.rs:432]: Ah, good point! I even saw that in the tests and amended the `mz_compute_error_counts` view accordingly. Will adjust the comment.

**Comment by antiguru** [comment]
> Also, record counts can be very misleading because a) errors are deduplicated and b) removed errors are only reflected after compaction:

I think this should be the expected behavior! Counting errors without _time_ might be confusing if at one time, there is a dataflow error, but at a different time, there isn't. Combining this with `AS OF` queries, at one time the query might succeed, but at a different time it won't. By only summing the error counts independently of time might convey the wrong message. However, from the arrangement batches we cannot determine which times are covered by what batch...

**Comment by teskje** [comment]
> I think this should be the expected behavior!

I don't think so. If I have three invalid records in an index or MV, I would expect the error count to be 3, not 1. If I then remove these three records I would expect the error count to become zero shortly after (with some introspection-induced delay), not become 2 and stay that for a while. I think that's independent of whether or not we log counts with times or not, since "1" and "2" are never correct answers for any time.

You are right that users might become confused if they use `AS OF` to see errors but introspection says there are no error counts because the errors have since been removed. But I would argue that:
 
* This is how all the other introspection sources work as well. They show you introspection data for the current time (or earlier times if you use `AS OF` in the introspection query), not all historical data with associated times.
* There is no reason that users should expect a query issued `AS OF` some previous time to be consistent with an introspection query issued `AS OF` now. Users could reasonably expect consistency when the introspection query was also issued `AS OF` the same previous time, but that would likely fail because we compact introspection arrangements aggressively.

In the logging operators we could easily also emit the current time, so it would be possible to log error counts with their respective times. But then this introspection source would just grow indefinitely.

The main use case for this is for users (and us) to quickly find out if there are errors in any dataflow. For that the current view is sufficient. If users want historical data, they could do as we do with Prometheus metrics and query periodically. Or, if we think that's too inconvenient, we should think about ways to configure the compaction windows of introspection sources.

File: src/compute/src/render/mod.rs [hunk 3]
```diff
@@ -613,6 +636,14 @@ where
                 // just create another handle to that arrangement.
                 let trace = compute_state.traces.get(&gid).unwrap().clone();
                 compute_state.traces.set(idx_id, trace);
+
+                // Announce error forwarding from the existing index.
+                if let Some(logger) = compute_state.compute_logger.clone() {
+                    logger.log(ComputeEvent::ErrorForward {
```
Comment (def-) [inline src/compute/src/render/mod.rs:642]: Interesting to cover?

File: src/compute/src/logging/compute.rs [hunk 4]
```diff
@@ -609,6 +687,47 @@ impl<A: Allocate> DemuxHandler<'_, '_, A> {
...
+            }
+        }
+    }
+
+    fn handle_error_forward(&mut self, from_export: GlobalId, to_export: GlobalId) {
+        // Log the initial error count.
+        if let Some(export) = self.state.exports.get(&from_export) {
+            self.handle_error_count(to_export, export.error_count);
+        } else {
+            error!(
```
Comment (def-) [inline src/compute/src/logging/compute.rs:713]: Possible to cover?

**Review by def- — COMMENTED** [review]
The SQLsmith errors from nightly look interesting, could they be related to this change? https://buildkite.com/materialize/nightlies/builds/4005

Marking interesting spots from coverage inline: https://buildkite.com/materialize/coverage/builds/227

File: src/adapter/src/catalog/builtin.rs [hunk 5]
```diff
@@ -5072,6 +5102,7 @@ pub static BUILTINS_STATIC: Lazy<Vec<Builtin<NameReference>>> = Lazy::new(|| {
         Builtin::Log(&MZ_COMPUTE_FRONTIERS_PER_WORKER),
         Builtin::Log(&MZ_COMPUTE_IMPORT_FRONTIERS_PER_WORKER),
         Builtin::Log(&MZ_COMPUTE_DELAYS_HISTOGRAM_RAW),
+        Builtin::Log(&MZ_COMPUTE_ERROR_COUNTS_RAW),
```
Comment (jkosh44) [inline src/adapter/src/catalog/builtin.rs:5105]: I think you're missing `MZ_ARRANGEMENT_RECORDS_RAW`.

Comment (umanwizard) [reply src/adapter/src/catalog/builtin.rs:5105]: That already exists, Jan just moved it around

File: src/compute/src/render/mod.rs (hunk 3 above)
Comment (teskje) [reply src/compute/src/render/mod.rs]: This part has been removed.

File: src/compute/src/logging/compute.rs (hunk 4 above)
Comment (teskje) [reply src/compute/src/logging/compute.rs]: This part has been removed. Wouldn't have been possible to cover because this was basically a "gentle assert" that would have alerted us to a bug without crashing the whole progress.

File: src/compute/src/logging/compute.rs [hunk 6]
```diff
@@ -373,19 +401,44 @@ struct DemuxState<A: Allocate> {
     peek_stash: BTreeMap<Uuid, Duration>,
     /// Arrangement size stash
     arrangement_size: BTreeMap<usize, ArrangementSizeState>,
+    /// Maps error-logging exports to exports they forward errors to.
+    error_forwards: BTreeMap<GlobalId, BTreeSet<GlobalId>>,
```
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute/src/logging/compute.rs]: This has since be removed, but my rationale was the it's possible to drop a dependency without dropping the dataflows that depend on it, and I wanted to keep the forwarding working even when the index that errors are forwarded from has already been dropped. Although that wouldn't have been strictly necessary, because the compute controller never drops dependencies when dependents still exist (but the replica code could handle that I think).

**Comment by teskje** [comment]
I rebased this PR with two major changes:

1. Remove the "error forwarding" logic in logging and instead put it into the `mz_compute_error_counts_per_worker` view. Turns out much simpler that way.
2. In logging, move the error counts from the diff field into the data field (see below).

> The SQLsmith errors from nightly look interesting, could they be related to this change?

Good catch! The issue was that per-worker error counts can be negative but we generally assume that the summed-up diffs are never negative, and violating that assumption makes various operators log errors. This is fixed now by (2) above.

**Comment by teskje** [comment]
Ready for another look!

**Review by jkosh44 — APPROVED** [review]
Adapter code LGTM

File: src/compute/src/logging/compute.rs [hunk 7]
```diff
@@ -890,3 +901,74 @@ impl Drop for RetractImportFrontiers {
...
+}
+
+impl<G, D> LogDataflowErrors for Collection<G, D, Diff>
+where
+    G: Scope,
+    D: Data,
+{
+    fn log_dataflow_errors(self, logger: Logger, export_id: GlobalId) -> Self {
+        self.inner
+            .unary(Pipeline, "LogDataflowErrors", |_cap, _info| {
```
Comment (antiguru) [inline src/compute/src/logging/compute.rs:995]: Nit: Rename to `LogDataflowErrorsCollection` and (`Stream`) below so the operators have unique names.

File: src/compute/src/logging/compute.rs [hunk 8]
```diff
@@ -890,3 +901,74 @@ impl Drop for RetractImportFrontiers {
...
+
+                    output.session(&cap).give_vec(&mut buffer);
+                });
+            }
+        })
+    }
+}
+
+/// Return the sum of all diffs within the given batch.
+fn sum_batch_diffs<B>(batch: &B) -> Diff
```
Comment (antiguru) [inline src/compute/src/logging/compute.rs:1035]: We might want to mention this is is `O(unique (value, key, time))`, so it can be slow for large collections. I don't think it's a problem since we only attach it to error streams, but we might want to add a note to our future selves that we don't attach it to non-error streams.

File: src/compute-client/src/logging.rs [hunk 9]
```diff
@@ -433,6 +436,11 @@ impl LogVariant {
             LogVariant::Compute(ComputeLog::ShutdownDuration) => RelationDesc::empty()
                 .with_column("worker_id", ScalarType::UInt64.nullable(false))
                 .with_column("duration_ns", ScalarType::UInt64.nullable(false)),
+
+            LogVariant::Compute(ComputeLog::ErrorCount) => RelationDesc::empty()
+                .with_column("export_id", ScalarType::String.nullable(false))
+                .with_column("worker_id", ScalarType::UInt64.nullable(false))
+                .with_column("count", ScalarType::Int64.nullable(false)),
```
Comment (antiguru) [inline src/compute-client/src/logging.rs:443]: Add the corresponding foreign keys to `LogVariant::foreign_key`.

File: src/compute/src/logging/compute.rs [hunk 10]
```diff
@@ -310,6 +313,14 @@ pub(super) fn construct<A: Allocate + 'static>(
             .as_collection()
             .map(arrangement_heap_datum_to_row);
 
+        let error_count = error_count.as_collection().map(move |datum| {
+            Row::pack_slice(&[
+                Datum::String(&datum.export_id.to_string()),
```
Comment (antiguru) [inline src/compute/src/logging/compute.rs:318]: We should have a different API that would allow us to not allocate a string here... Not for this PR, let's leave it as follow-up work.

File: src/compute/src/logging/compute.rs [hunk 11]
```diff
@@ -310,6 +313,14 @@ pub(super) fn construct<A: Allocate + 'static>(
...
+                Datum::UInt64(u64::cast_from(worker_id)),
+                Datum::Int64(datum.count),
```
Comment (antiguru) [inline src/compute/src/logging/compute.rs:320]: Can we move the `count` into the diff of the collection?

File: src/compute/src/logging/compute.rs [hunk 12]
```diff
@@ -618,6 +665,36 @@ impl<A: Allocate> DemuxHandler<'_, '_, A> {
...
+        }
+        if new_count != 0 {
+            let datum = ErrorCountDatum {
+                export_id,
+                count: new_count,
+            };
+            self.output.error_count.give((datum, ts, 1));
+        }
+
+        export.error_count = new_count;
```
Comment (antiguru) [inline src/compute/src/logging/compute.rs:695]: This would go away if we use the diff as a differential diff.

File: src/catalog/src/builtin.rs [hunk 13]
```diff
@@ -3245,8 +3245,33 @@ pub const MZ_COMPUTE_ERROR_COUNTS_PER_WORKER: BuiltinView = BuiltinView {
...
+        FROM mz_internal.mz_compute_error_counts_raw
+    ),
+    -- Error counts that were not directly logged but forwarded from reused indexes.
+    indirect_errors AS (
+        SELECT r.object_id, e.worker_id, e.count
+        FROM index_reuses r
+        JOIN direct_errors e
+            ON (r.dependency_id = e.export_id)
+    )
+SELECT * FROM direct_errors UNION ALL (SELECT * FROM indirect_errors)",
```
Comment (antiguru) [inline src/catalog/src/builtin.rs:3274]: It's great to have the error count based on the transitive closure, but I have two concerns:
1. Does `index_reuses` recursively show indexes? I don't have the semantics of `mz_compute_dependencies/exports` in my head, so this might be the case already! I think we need to use the transitive closure here.
2. It feels this should be a separate view, because otherwise it might be more difficult to pin where errors come from, but I don't have a strong opinion.

File: src/compute/src/logging/compute.rs [hunk 14]
```diff
@@ -890,3 +901,74 @@ impl Drop for RetractImportFrontiers {
...
+
+                        output.session(&cap).give_vec(&mut buffer);
+                    });
+                }
+            })
+            .as_collection()
+    }
+}
+
+impl<G, B> LogDataflowErrors for Stream<G, B>
```
Comment (antiguru) [inline src/compute/src/logging/compute.rs:1012]: Do we need this implementation? The arrangement that we're reading from already logs the counts, so we could re-use this without needing to count again. It sure makes the queries more complicated, but would not incur the cost of `sum_batch_diffs`.

File: src/compute/src/logging/compute.rs (hunk 7 above)
Comment (teskje) [reply src/compute/src/logging/compute.rs:995]: We often have this pattern that the same operator can be attached to different object types (streams, collections, arrangements) via an extension trait. Are you suggesting that it not a good pattern and we should move away from it?

Fine with doing that in this case, but then I wouldn't bother having traits in the mix at all and just implement free-standing `log_collection_dataflow_errors` and `log_stream_dataflow_errors` functions.

Comment (antiguru) [reply src/compute/src/logging/compute.rs:995]: No, that pattern is fine -- that's the purpose to having a trait here. I'm thinking more about debugging where we could pinpoint the implementation from the name, instead of having to go back to the callsite to determine which implementation we might select.

File: src/compute/src/logging/compute.rs (hunk 14 above)
Comment (teskje) [reply src/compute/src/logging/compute.rs:1012]: > The arrangement that we're reading from already logs the counts

You mean the arrangement size logging, right? That logs the number of physical records, not the logical number of errors. That count can be wildly different if the arrangement isn't fully compacted, so I don't think we can use it. See my comment above for an example that shows how misleading that would be.

File: src/compute/src/logging/compute.rs (hunk 7 above)
Comment (teskje) [reply src/compute/src/logging/compute.rs:995]: Oh, you meant the visible name, not the trait name! Sorry :D

File: src/compute/src/logging/compute.rs (hunk 11 above)
Comment (teskje) [reply src/compute/src/logging/compute.rs:320]: It was like that before, but I changed it in response to the invalid retraction errors seen in CI: https://github.com/MaterializeInc/materialize/pull/21640#pullrequestreview-1625514879.

We assume that collections exported by compute always sum up to non-negative diffs. But per-worker error counts can be negative (as you pointed out elsewhere), so we can't encode them as diffs. Unless I'm missing something?

Comment (antiguru) [reply src/compute/src/logging/compute.rs:320]: Good point; we might be able to use negative counts if we have a `_RAW` collection that contains the negative counts, and a view on top that does a `COUNT(*)`, which should reveal the negative counts as data. Ofc, this might break down somewhere on the way.

File: src/catalog/src/builtin.rs (hunk 13 above)
Comment (teskje) [reply src/catalog/src/builtin.rs:3274]: 1. No, it's not recursive. This view is based on the assumption that indexes are never transitively reused. I.e. if there is one dataflow producing an arrangements, copies of that dataflow will all depend on the first one, not on each other. I only did a cursory test to verify this, which is not enough. I'll verify in the code.
2. Hm, I would like to avoid users having to know about the fact that we reuse indexes. I don't think this makes it more difficult to find errors either! If we report an error in a reused index, you can look at our various dependency views to check if any dependencies have errors. Or you can look at the SQL of the view's index view `SHOW CREATE`, or you can look at the `EXPLAIN` for that index. Each of these work the same for reused and non-reused indexes.

File: src/compute/src/logging/compute.rs (hunk 11 above)
Comment (teskje) [reply src/compute/src/logging/compute.rs:320]: Yep, that was what the previous implementation was doing. The issue is, I think, that we provide a `_per_worker` view, which groups and sums counts per-worker, so the result can be negative.

We the the `_per_worker` view mainly for consistency (we have it for all the other introspection sources)`. We could skip it and only have a view that aggregates over all workers. This might work. However, we would be relying on error semantics and those are pretty weird right now. Also, we would still have queryable `_raw` collections that still can give you an error and that might still cause confusion down the road.

Comment (antiguru) [reply src/compute/src/logging/compute.rs:320]: Alright, let's leave it as-is. This is not a high-volume introspection source, so it should be OK. We could put a worker-local arrangement in front if it to avoid allocations that would cancel out, but you decide if it'd be worth it.

File: src/compute-client/src/logging.rs (hunk 9 above)
Comment (teskje) [reply src/compute-client/src/logging.rs:443]: A bunch of other foreign keys are also missing and it is not clear how declaring them is useful. We should either:

* Do a pass to make sure all foreign keys are correctly and completely defined or
* Stop declaring foreign keys if it's not useful.

In any case I think we should do that separately from this PR. [I've started a discussion in Slack.](https://materializeinc.slack.com/archives/CM7ATT65S/p1696589788239279)

File: src/compute/src/logging/compute.rs (hunk 11 above)
Comment (teskje) [reply src/compute/src/logging/compute.rs:320]: I decided to not do this here as it seems inconsistent: Even in the worst case, the error count of a dataflow should update roughly as frequent as the dataflow frontier. But we don't do a pre-arrange for frontiers, so it's difficult to justify doing one for error counts. Maybe we should do a pre-arrange for both (and other logging arrangements too), but that's outside the scope of this PR.

File: src/compute/src/logging/compute.rs (hunk 10 above)
Comment (teskje) [reply src/compute/src/logging/compute.rs:318]: How would that look like? Having a dedicated `Datum::GlobalId` variant?

**Comment by antiguru** [comment]
Related to what we talk about yesterday, could you check how much time the dataflow spends in processing error counts versus the rest of what it's doing? It should be very little, and we don't expect many errors, but it'd be great to avoid surprises!

File: src/compute/src/logging/compute.rs (hunk 10 above)
Comment (antiguru) [reply src/compute/src/logging/compute.rs:318]: No, simpler than that. Something like this:
```
error_count.as_collection().map_batch(move |container| {
	let mut string = String::new();
    container.map(move |datum| {
		string.clear();
		write!(&mut string, "{}", datum.export_id);
		Row::pack_slice(&[Datum::String(&string), ...])
  	})
});
```

File: src/catalog/src/builtin.rs (hunk 13 above)
Comment (teskje) [reply src/catalog/src/builtin.rs]: > This view is based on the assumption that indexes are never transitively reused.

AFAICT that's only the case today because index imports get sorted by the index ID so the oldest index is always the first and that's the one that gets then picked. This is an implementation detail we should rely on, so I made this view check transitive reuses instead.

File: src/compute/src/logging/compute.rs (hunk 10 above)
Comment (teskje) [reply src/compute/src/logging/compute.rs:318]: Ah! But why do we need a new API for that? This should work:

```rust
        let mut export_id = String::new();
        let error_count = error_count.as_collection().map(move |datum| {
            export_id.clear();
            write!(&mut export_id, "{}", datum.export_id).unwrap();
            Row::pack_slice(&[
                Datum::String(&export_id),
                Datum::UInt64(u64::cast_from(worker_id)),
                Datum::Int64(datum.count),
            ])
        });
```

Comment (teskje) [reply src/compute/src/logging/compute.rs:318]: Opened a separate PR to make this optimization for every log variant: https://github.com/MaterializeInc/materialize/pull/22236

**Comment by teskje** [comment]
I've run two tests to see how much work the error logging operators do. The first test didn't introduce any dataflow errors, the second one had a dataflow with continually changing error counts.

## Test 1 (no errors)

### Setup
* TPCH load generator (SF 1, tick interval 1s)
* 2xsmall compute replica with INTROSPECTION DEBUGGING enabled
* index on TPCH query 20

### Record counts in introspection arrangements

```
                       name                       | records
--------------------------------------------------+---------
 ArrangeByKey Compute(ErrorCount)                 |       0
 ArrangeByKey Compute(PeekCurrent)                |       2
 ArrangeByKey Compute(PeekDuration)               |      13
 ArrangeByKey Compute(FrontierDelay)              |      60
 ArrangeByKey Compute(DataflowCurrent)            |       3
 ArrangeByKey Compute(FrontierCurrent)            |      78
 ArrangeByKey Compute(ShutdownDuration)           |      16
 ArrangeByKey Compute(ArrangementHeapSize)        |      47
 ArrangeByKey Compute(ImportFrontierCurrent)      |      15
 ArrangeByKey Compute(ArrangementHeapCapacity)    |      47
 ArrangeByKey Compute(ArrangementHeapAllocations) |      47
```

As expected, there are no records in the `ErrorCount` arrangement.

### Operator elapsed time

* Whole `compute logging` dataflow: 1315182217ns
* Error count logging operators: 6430474ns

The absolute numbers above are not very interesting as they depend on how long after the replica creation the measurement was taken. But they show that the portion of time spent computing the error counts is small:

* `6430474 / 1315182217 ~ 0.5%`

### Messages sent

The number of messages sent to the error count logging operators is zero, as expected.

## Test 2 (continually changing error counts)

### Setup
* TPCH load generator (SF 1, tick interval 1s)
* 2xsmall compute replica with INTROSPECTION DEBUGGING enabled
* index on a modified version of TPCH query 20 that always contains some amount of errors:

```sql
CREATE VIEW q20_errors AS
SELECT s_name, s_address
FROM supplier, nation
WHERE 
  s_suppkey IN (
    SELECT ps_suppkey
    FROM partsupp
    WHERE
      ps_partkey IN (SELECT p_partkey FROM part WHERE p_name like 'forest%')
      AND ps_availqty > (
        SELECT 0.5 * sum(l_quantity) / (sum(counter) % sum(l_quantity))
        FROM lineitem, count
        WHERE
          l_partkey = ps_partkey
          AND l_suppkey = ps_suppkey
          AND l_shipdate >= DATE '1995-01-01'
          AND l_shipdate < DATE '1995-01-01' + INTERVAL '1' year))
  AND s_nationkey = n_nationkey
  AND n_name = 'CANADA'
ORDER BY s_name;
``` 

`count` is a COUNTER load generator and the division by `sum(counter) % sum(l_quantity)` introduces a continually changing amount of division-by-zero errors as both the COUNTER and the TPCH load generators tick.

### Record counts in introspection arrangements

```
                       name                       | records
--------------------------------------------------+---------
 ArrangeByKey Compute(ErrorCount)                 |       3
 ArrangeByKey Compute(PeekCurrent)                |       2
 ArrangeByKey Compute(PeekDuration)               |      16
 ArrangeByKey Compute(FrontierDelay)              |      60
 ArrangeByKey Compute(DataflowCurrent)            |       3
 ArrangeByKey Compute(FrontierCurrent)            |      78
 ArrangeByKey Compute(ShutdownDuration)           |      12
 ArrangeByKey Compute(ArrangementHeapSize)        |      30
 ArrangeByKey Compute(ImportFrontierCurrent)      |      18
 ArrangeByKey Compute(ArrangementHeapCapacity)    |      30
 ArrangeByKey Compute(ArrangementHeapAllocations) |      28
(11 rows)
```

### Operator elapsed time

* Whole `compute logging` dataflow: 579347468ns
* Error count logging operators: 15212764ns

Portion of time spent computing the error counts:

* `15212764 / 579347468 ~ 2.6%`

### Messages sent

* Messages sent within the whole `compute logging` dataflow: 591909
* Messages sent to error count logging operators: 6591

Portion of messages sent to error count logging operators:

* `6591 / 591909 ~ 1.1%`

## Summary

When no errors are present the overhead of the error logging operators is not noticeable.

When error counts are quickly changing, the overhead of the error logging operators is noticeable but still small.

**Review by antiguru — APPROVED** [review]
Thanks for the thorough performance evaluation! Only a question: there only appear 3 records in the error arrangement? This shouldn't stop merging the change, just curious why that is. (I didn't again look at the implementation)

**Comment by teskje** [comment]
> there only appear 3 records in the error arrangement?

This is my guess: There is only a single dataflow with errors, so if the `ErrorCount` arrangement is fully compacted, it should only contain a single record. That it contains 3 means it isn't fully logically compacted, probably because of https://github.com/MaterializeInc/materialize/issues/11138#issuecomment-1684320014. So the records are probably:
* `(<old count>, ts1, +1)`
* `(<old count>, ts2, -1)`
* `(<new count>, ts2, +s)`
